sdk/identity/azure-identity/azure/identity/_internal/msal_credentials.py

```diff
@@ -57,21 +94,53 @@ def _get_app(self):
         # type: () -> msal.ClientApplication
         pass
 
-    def _create_app(self, cls):
-        # type: (Type[msal.ClientApplication]) -> msal.ClientApplication
+    def _create_app(self, cls, **kwargs):
+        # type: (Type[msal.ClientApplication], **Any) -> msal.ClientApplication
         """Creates an MSAL application, patching msal.authority to use an azure-core pipeline during tenant discovery"""
 
         # MSAL application initializers use msal.authority to send AAD tenant discovery requests
         with self._adapter:
             # MSAL's "authority" is a URL e.g. https://login.microsoftonline.com/common
-            app = cls(client_id=self._client_id, client_credential=self._client_credential, authority=self._base_url)
+            app = cls(
+                client_id=self._client_id,
+                client_credential=self._client_credential,
+                authority=self._base_url,
+                token_cache=self._cache,
+                **kwargs
+            )
 
         # monkeypatch the app to replace requests.Session with MsalTransportAdapter
         app.client.session.close()
         app.client.session = self._adapter
 
         return app
 
+    def _account_matches_profile(self, account):
+        return (
+            account.get("home_account_id") == self._profile.home_account_id
+            and account.get("username") == self._profile.username
+            and account.get("environment") == self._profile.environment
+            and account.get("realm") == self._profile.tenant_id
+        )
+
+    def _acquire_token_silent(self, *scopes, **kwargs):
+        if self._profile:
+            app = self._get_app()
+            accounts = [
+                acct
+                for acct in app.get_accounts(username=self._profile.username)
+                if self._account_matches_profile(acct)
+            ]
+            for account in accounts:
+                # MSAL asserts scopes is a list
+                scopes = list(scopes)  # type: ignore
+                now = int(time.time())
+                token = app.acquire_token_silent(scopes, account=account, **kwargs)
+                # TODO: could cache account
+                if token and "access_token" in token and "expires_in" in token:
+                    return AccessToken(token["access_token"], now + int(token["expires_in"]))
```

Comment:
>return AccessToken(token["access_token"], now + int(token["expires_in"])) [](start = 20, length = 73)

can you add  token_type in AccessToken which is needed for client to use, by default token_type is "Bearer"

Reply:
According to the documentation I've read ([for example](https://docs.microsoft.com/en-us/azure/active-directory/azuread-dev/v1-protocols-oauth-code#successful-response-1)) Azure AD provides only bearer tokens in the flows we use. What other types do you expect, and how does the client use them?

Reply:
we add this type in the authorization head of http request. AAD may add new type in the future so that I would prefer not hard-code in client code. Currently CLI get the value from ADAL token information.

---
In reply to: [398762506](https://github.com/chlowell/azure-sdk-for-python/pull/4#discussion_r398762506) [](ancestors = 398762506)

Reply:
I don't want to extend `AccessToken` without a definite use case. If AAD adds a new token type in the future, using it in azure-identity will require design and implementation work I don't want to try to preempt today.